src/libraries/System.Text.Json/src/System/Text/Json/Serialization/ReadStackFrame.cs

```diff
@@ -126,22 +127,26 @@ internal void InitializeRequiredPropertiesValidationState(JsonTypeInfo typeInfo)
         {
             Debug.Assert(RequiredPropertiesLeft == null);
 
-            if (typeInfo.RequiredPropertiesIndices != null)
+            if (typeInfo.NumberOfRequiredProperties > 0)
             {
-                RequiredPropertiesLeft = new HashSet<int>(typeInfo.RequiredPropertiesIndices);
+                RequiredPropertiesLeft = new BitArray(typeInfo.NumberOfRequiredProperties, defaultValue: true);
             }
         }
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         internal void ValidateAllRequiredPropertiesAreRead(JsonTypeInfo typeInfo)
         {
-            if (typeInfo.RequiredPropertiesIndices != null)
+            if (typeInfo.NumberOfRequiredProperties > 0)
             {
                 Debug.Assert(RequiredPropertiesLeft != null);
 
-                if (RequiredPropertiesLeft.Count != 0)
+                // Optimize this when https://github.com/dotnet/runtime/issues/72999 is fixed
+                for (int i = 0; i < RequiredPropertiesLeft.Count; i++)
```

Comment:
Given that this now contains a for loop, I would remove `AggressiveInlining`. You might want to consider writing `BitArray.Contains` as an extension method so that this method body remains small.

Reply:
I'll leave ValidateAllRequiredPropertiesAreRead as inlined so that we don't do call if `if` check fails which is most common but will add Contains.